### PR TITLE
[PF-504] Add application role to WSM

### PR DESCRIPTION
**Pull request by zloery**
(Part 1 of 4? changes for controlled resource IAM in WSM)

This change adds the new `APPLICATION` workspace IAM role to WSM. The full set of workspace roles is now `READER`, `WRITER`, `APPLICATION`, `OWNER`. This is  used for managing controlled resources, and was added to Sam in https://github.com/broadinstitute/sam/pull/509

(This doesn't add the `EDITOR` role, which we decided not to use. `WRITERS` will have permission to create/delete resources instead)

APPLICATION is a role granted to registered applications in a workspace. It is equivalent to editor but can also create application-controlled resources. Unlike other roles, this is only granted through application registration (not yet implemented), rather than through WSM's role management endpoints. For now, it has the same GCP project-level permissions as writers and owners.

### CODE REVIEW

File: src/test/java/bio/terra/workspace/service/iam/SamServiceTest.java [hunk 1]
```diff
@@ -171,12 +171,18 @@ void ListPermissionsIncludesAddedUsers() {
...
+        RoleBinding.builder().role(IamRole.APPLICATION).users(Collections.emptyList()).build();
+    RoleBinding expectedEditorBinding =
+        RoleBinding.builder().role(IamRole.EDITOR).users(Collections.emptyList()).build();
     assertThat(
         policyList,
         containsInAnyOrder(
             equalTo(expectedOwnerBinding),
             equalTo(expectedWriterBinding),
-            equalTo(expectedReaderBinding)));
+            equalTo(expectedReaderBinding),
```
Comment (jaycarlton) [inline src/test/java/bio/terra/workspace/service/iam/SamServiceTest.java:183]: interesting. I thought you could make a list of the expected values and plug it in there.

**Review by jaycarlton — APPROVED** [review]

File: src/main/java/bio/terra/workspace/service/iam/model/IamRole.java [hunk 2]
```diff
@@ -1,12 +1,15 @@
...
 
+import bio.terra.workspace.service.iam.exception.InvalidRoleException;
 import java.util.Arrays;
 import java.util.Optional;
 
 /** Internal representation of IAM roles. */
 public enum IamRole {
   READER("reader", bio.terra.workspace.generated.model.IamRole.READER),
   WRITER("writer", bio.terra.workspace.generated.model.IamRole.WRITER),
+  EDITOR("editor", bio.terra.workspace.generated.model.IamRole.EDITOR),
```
[the remark this answers is not in the document]
Comment (zloery) [reply src/main/java/bio/terra/workspace/service/iam/model/IamRole.java:11]: Yep, though I think we want to change all the API objects at once.

File: src/main/java/bio/terra/workspace/service/workspace/CloudSyncRoleMapping.java [hunk 3]
```diff
@@ -6,37 +6,31 @@
...
-                  "roles/lifesciences.editor",
-                  "roles/serviceusage.serviceUsageConsumer",
-                  // TODO(wchambers): Revise notebooks permissions when there are controlled
-                  // resources for notebooks.
-                  "roles/notebooks.admin",
-                  // TODO(marikomedlock): Revise storage permissions when there are controlled
-                  // resources for buckets.
-                  "roles/storage.admin"),
-          IamRole.READER, ImmutableList.of("roles/viewer"));
+          IamRole.OWNER, WRITER_PERMISSIONS,
```
[the remark this answers is not in the document]
Comment (zloery) [reply src/main/java/bio/terra/workspace/service/workspace/CloudSyncRoleMapping.java:32]: done

File: src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java [hunk 4]
```diff
@@ -62,11 +62,18 @@ public StepResult doStep(FlightContext flightContext)
           "group:" + workingMap.get(WorkspaceFlightMapKeys.IAM_READER_GROUP_EMAIL, String.class);
       String writerGroup =
           "group:" + workingMap.get(WorkspaceFlightMapKeys.IAM_WRITER_GROUP_EMAIL, String.class);
+      String applicationGroup =
+          "group:"
```
[the remark this answers is not in the document]
Comment (zloery) [reply src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java]: done

File: src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java [hunk 5]
```diff
@@ -94,7 +101,7 @@ public StepResult doStep(FlightContext flightContext)
    */
   private List<Binding> bindingsForRole(IamRole role, String group) {
     List<Binding> bindings = new ArrayList<>();
-    for (String gcpRole : CloudSyncRoleMapping.cloudSyncRoleMap.get(role)) {
+    for (String gcpRole : CloudSyncRoleMapping.CLOUD_SYNC_ROLE_MAP.get(role)) {
```
[the remark this answers is not in the document]
Comment (zloery) [reply src/main/java/bio/terra/workspace/service/workspace/flight/GoogleCloudSyncStep.java]: done

File: src/main/java/bio/terra/workspace/service/iam/SamService.java [hunk 6]
```diff
@@ -292,6 +292,9 @@ public SystemStatusSystems status() {
         new AccessPolicyMembership()
             .addRolesItem(IamRole.OWNER.toSamRole())
             .addMemberEmailsItem(ownerEmail));
+    policyMap.put(
```
Comment (ddietterich) [inline src/main/java/bio/terra/workspace/service/iam/SamService.java:295]: Maybe update the method comment

**Review by ddietterich — APPROVED** [review]
LGTM

**Pull request merged**